### PR TITLE
cloud: schema feature added

### LINKED ISSUE #372: cloud: Support specifying the database schema to deploy
At the moment the [cloud tool](https://github.com/kernelci/kcidb/blob/main/cloud) doesn't specify the version of the schema to initialize the databases to, when deploying, so the databases get initialized to the latest version. Add support for specifying the schema version using a command-line option with the `cloud deploy` command. E.g. `--schema VERSION`, similar to the `kcidb-db-init` tool used for actual initialization.

This will make it easier to make deployments with older schema versions, for testing upgrades, and reproducing issues with existing (older-schema) deployments.

### PR DESCRIPTION
**Pull request by octonawish-akcodes**
Support for schema version for initializing the db with versions specified.

This PR will closes #372 

@spbnick need reviews

Am I on the right track? I would appreciate some guidance. Also how can I test it.

### CODE REVIEW

File: cloud [hunk 1]
```diff
@@ -1260,6 +1265,7 @@ function cloud_functions_env() {
                           database \
                           clean_test_databases \
                           empty_test_databases \
+                          schema \
```
Comment (spbnick) [inline cloud:1268]: I don't think we need to pass the schema version to the environment. We could discover it from the databases themselves, if needed.

File: cloud [hunk 2]
```diff
@@ -1675,6 +1681,7 @@ function execute_command() {
                           updated_publish \
                           submitters \
                           argv \
+                          schema \
```
Comment (spbnick) [inline cloud:1684]: Let's keep the variable name consistent throughout the code, if we can. I think `schema_version` works well for that.

**Review by spbnick — CHANGES_REQUESTED** [review]
This is a good start, now you need to write the code to accept (and validate) the version number from the command line, pass it around where it's needed, and apply it to other database types as well.

Regarding testing, I've created a Google Cloud project for you to test your changes. It's called `kcidb-abhishek` and you should've received an invitation to join it via email. Next you need to install the Google Cloud SDK, create a service account with administrator permissions, and create and download a key for it. Then you could export its location via the `GOOGLE_APPLICATION_CREDENTIALS` environment variable, and after that you would be able to run the cloud tool to make a test deployment. Try with an unmodified `cloud` tool first. The command to deploy to your project would look something like `./cloud deploy kcidb-abhishek "" 1 -v --smtp-mocked --test --heavy-asserts`. Use the [Administrator Guide](https://kernelci.org/docs/kcidb/administrator_guide/) as a reference.

File: cloud (hunk 1 above)
Comment (octonawish-akcodes) [reply cloud]: Okay

File: cloud (hunk 2 above)
Comment (octonawish-akcodes) [reply cloud]: okay

**Comment by octonawish-akcodes** [comment]
@spbnick Need reviews

**Review by spbnick — CHANGES_REQUESTED** [review]
This is starting to look good. Thank you, @octonawish-akcodes. I left a few comments inline.

However, I just realized that we can't just pass the same schema version to each database type (BigQuery/PostgreSQL/sqlite), as they could be totally different, because the schema development for different databases has different needs and moves at different speeds.

However, we can specify the I/O schema version, which the databases should support instead, and that has to be the same. The `kcidb.db.Client` class [supports](https://github.com/kernelci/kcidb/blob/cfcfe8f2d7940f6b0ae7ebe15d6a19ffaca17ea2/kcidb/db/__init__.py#L96-L101) that, but none of the database-managing command-line tools do. So we would need a commit before this one adding support for that.

We could make this work if we e.g. allow the `--schema` options for these tools to accept a string in the shape of `io:<major>.<minor>` for the I/O schemas. While we're at it, we could also allow it to accept the string `latest`, meaning the latest database schema version, and that would help you get rid of those "specified/unspecified schema version" branches you're adding to database initialization functions in `cloud`.

**Comment by octonawish-akcodes** [comment]
Tried to implement I/O versions to the db command-line tools, am I on the right track? Need reviews.

File: cloud [hunk 3]
```diff
@@ -2218,6 +2241,7 @@ function execute() {
             --heavy-asserts) heavy_asserts="true"; shift;;
             --mute-updates) updated_publish="false"; shift;;
             --test) test="true"; shift;;
+            --schema) schema_version="$2"; shift 2;;
```
[the remark this answers is not in the document]
Comment (octonawish-akcodes) [reply cloud]: Okay

File: cloud [hunk 4]
```diff
@@ -2164,7 +2186,7 @@ function execute() {
         getopt_longopts+=",smtp-mocked,test"
         if [[ $command == @(deploy|withdraw) ]]; then
             getopt_shortopts+="vs:"
-            getopt_longopts+=",verbose,sections:,submitter:"
+            getopt_longopts+=",verbose,sections:,submitter:,schema:"
```
[the remark this answers is not in the document]
Comment (octonawish-akcodes) [reply cloud]: But like in the withdrawing stage doesnt it need the schema values?

[the remark this answers is not in the document]
Comment (spbnick) [reply cloud]: The "withdraw" and all other stages would be able to retrieve the schema version from the databases themselves, so only "deploy" needs it to know which schema to initialize the databases to.

**Review by spbnick — CHANGES_REQUESTED** [review]
This is a good start.

Note that only two tools use the `--schema` option for now: `kcidb-db-init` and `kcidb-db-upgrade`, no other tool needs it (including `kcidb-db-query`).

Instead of copying the option-parsing code around, create another "argparse type" function, similar to `kcidb.misc.version()` and `schema_version()` inside `kcidb.misc.argparse_schema_add_args()`. Then both the tools would be able to use it. The function should produce a value which would be accepted by both `kcidb.db.Client.init()` and `kcidb.db.Client.upgrade()`.

Actually, this might need an extraction of a method looking up a database schema corresponding to an I/O schema, from both `init()` and `upgrade()`, in `kcidb.db.Client`, and then use of that method in both of them. This would make it possible to use that new method to check and report an error whenever the resulting database version is older than the current one, in `kcidb-db-upgrade` (instead of just failing an assertion in `init()` and `upgrade()`).

**Pull request closed without merging**